### PR TITLE
SparseConv3D Mixed Precision Training Support

### PR DESCRIPTION
**Pull request by CCInc**
I added support for mixed-precision training in torchsparse (https://github.com/mit-han-lab/torchsparse/pull/69) with the following results:

> The performance of the mixed-precision training is almost the same as that of the full-precision training: 76.43 v.s. 76.78.
> The speedup is fairly limited: the total training time is reduced from 7.3 hours to 6.4 hours (around 10% reduction).
> The memory reduction is very significant: the memory usage is reduced from 48.8G to 28.8G (40% reduction).

With these commits, I added a `training.enable_mixed` flag which enables the usage of mixed precision on supported models, as well as a `_supports_mixed` variable to the models which support mixed precision (sp3d with torchsparse backend). 

To actually activate mixed precision, I wrapped the training in `torch.cuda.amp.autocast` and added a loss scaler `torch.cuda.amp.GradScaler`.

I'm running experiments on s3dis which you can find here: https://wandb.ai/allcowseatfood/s3dis 
You can see the mixed precision training has lower memory usage than regular training and autocasted minkowskiengine. 

image

### CODE REVIEW

File: conf/training/minkowski_scannet.yaml [hunk 1]
```diff
@@ -1,11 +1,12 @@
+# @package _global_
 # Ref: https://github.com/chrischoy/SpatioTemporalSegmentation/blob/master/config.py
-# @package training
```
Comment (nicolas-chaulet) [inline conf/training/minkowski_scannet.yaml:2]: Is this intentional? I thought that this should be `training`

**Review by nicolas-chaulet — COMMENTED** [review]
Looks great, thanks @CCInc !! 
Could you please add a note in the README maybe under the sparse conv section?
Also, add an entry to the changelog.

**Comment by nicolas-chaulet** [comment]
@CCInc looks like the linter is not super happy, I'll let you fix that and update the doc before approving

**Comment by CCInc** [comment]
Yeah... not sure what to do about that without reworking that whole thing, I think I'll just add `global` to that var so it passes the linter for now.

File: conf/training/minkowski_scannet.yaml (hunk 1 above)
Comment (CCInc) [reply conf/training/minkowski_scannet.yaml:2]: It's kind of broken, `wandb` needs to be in the global namespace, which you can't do with `@package training`. I'll be fixing all this with #614

**Comment by nicolas-chaulet** [comment]
Can't you just deactivate for that line? Adding a global might not be a bad thing, that's what it is in the end...

**Comment by humanpose1** [comment]
That's great ! thanks a lot for your contribution @CCInc !
My data have a lot of points, I'm looking forward to testing mixed precision.
I have a question,  mixed precision is just for training step or it can also be applied for validation or test step (if we want to be faster or less memory footprint) ?

**Comment by CCInc** [comment]
@humanpose1 It should be supported for training and inference, you can also test it by looking at the dtype of the output of the backbone in your models:

```
        features = self.backbone(self.input).x
        print(features.dtype)
```

It should reduce memory footprint a good amount in inference. But, usually from what I see sparseconv networks are usually fully saturated on the computing, so increasing batch size won't get much speed gain (I haven't tried it on anything newer than a Titan X though so it maybe would!)

**Review by nicolas-chaulet — APPROVED** [review]
Great, thanks a lot @CCInc !
